### ISSUE TITLE
Load balacing

### ISSUE THREAD

**Issue by florian-lefebvre** [op-ent/api#9]
J'y connais rien mais il faut le mettre en place rapidement. Toute ressource serait appréciée, hésitez pas à mettre des liens dans la conv.

**Comment by florian-lefebvre** [comment]
@SkanixDev On peut utiliser pm2 avec `pm2 start server.js -i max` (https://pm2.keymetrics.io/docs/usage/cluster-mode/), qu'est-ce que tu en penses ?

**Comment by SkanixDev** [comment]
Yep ! Pour mes projets perso  j'ai toujours utilisé pm2 et je trouve super pratique donc ouais je suis pour après peut être que il y a d'autre technologie mieux ?

**Comment by florian-lefebvre** [comment]
Je pense pas mais je voudrais l'avis de @MRakox

**Comment by MRakox** [comment]
Si j'ai bien compris, le mode cluster de pm2 sert à utiliser une application en multi-thread. Autrement dit, c'est pour utiliser tous les CPUs / les ressources disponibles.
Sur le principe, le load balancing se fait plutôt sur plusieurs machines. Plusieurs instances de l'application sont lancées en même temps et un proxy / un service (comme CF) distribue les requêtes entre les différents nodes. 
La principale contrainte technique, à ce niveau là, il faut que les différentes instances renvoient les mêmes informations, il faut qu'elles soient tout le temps synchronisées. Ça impacte notamment les technologies choisies pour le stockage de données en cache (pas de stockage en mémoire, privilégier Redis ou d'autres solutions partagées) ou encore pour les jobs (il faut qu'ils soient exécutés uniquement sur un seul node pour éviter les doublons).

**Comment by florian-lefebvre** [comment]
Je vois. C'est quelque chose qui se fait avec NGINX non ? Et est-ce que tu as des exemples de setup (repo ou quelques codes snippets) ?

**Comment by MRakox** [comment]
Quelques explications sur le principe du load balancing :
https://aws.amazon.com/what-is/load-balancing/

**Comment by MRakox** [comment]
Un exemple avec AWS :
https://adevait.com/adonisjs/how-to-deploy-adonisjs-application-with-aws-elastic-beanstalk
Un exemple avec Cloudflare : https://docs.cleavr.io/guides/load-balancer/

**Comment by florian-lefebvre** [comment]
Super merci !

**Comment by MRakox** [comment]
Maintenant, ça sera peut être pas utile au début du projet d'avoir plusieurs instances parallèles.
Mais c'est toujours bon d'y réfléchir assez tôt pas se retrouver embêter quand ça devient nécessaire pour augmenter les performances de l'application...

**Comment by florian-lefebvre** [comment]
Clairement 👍. Je pense que dans un 1er temps on va juste utiliser pm2 et quand ça deviendra trop chaud (c'est pas pour de suite), on utilisera une solution de load balacing avec cloudflare par exemple

**Comment by Micorksen** [comment]
Je pensais qu'il fallait plusieurs machines avec les mêmes processus, CloudFlare fait du load balancing ?

**Comment by MRakox** [comment]
En effet, il faut bien plusieurs machines avec la même application pour mettre en place une solution de load balancing. Cloudflare se charge uniquement de la partie proxy ainsi que de repartir équitablement le traffic entre les différentes instances.

**Comment by florian-lefebvre** [comment]
Donc le but serait de payer plusieurs VPS (chez OVH ou droplets sur Digital Ocean par exemple mais ça change pas grand chose) et de gérer le load balacing + la protection DDOS + le domain name chez cloudflare, qu'est-ce que vous en pensez ?

**Comment by MRakox** [comment]
Ça me paraît bien !
Je suis pas un grand fan des solutions all in one à la Google Cloud / Amazon Web Services...

**Comment by florian-lefebvre** [comment]
Moi non plus 👍

**Comment by florian-lefebvre** [comment]
Pour la protection DDOS et le domain name chez cloudflare, faudra attendre quelques mois pour que je puisse le transférer depuis OVH mais c'est pas un souci je pense vu la taille du projet pour le moment

**Comment by Micorksen** [comment]
> Pour la protection DDOS et le domain name chez cloudflare, faudra attendre quelques mois pour que je puisse le transférer depuis OVH mais c'est pas un souci je pense vu la taille du projet pour le moment

Mets simplement les DNS Cloudflare, le load balancing se fait avec un champs custom LB, pas besoin de transfert. Et puis pourquoi tu dois attendre? Normalement si t'as un code d'auth EEP tu peux sans attendre.

**Comment by florian-lefebvre** [comment]
J'ai pas encore essayé mais de ce que j'ai lu c'est régulé par l'ICANN. Je vais pas pas regarder de suite parce que on en a pas vraiment besoin
